File: examples/example_mcap_writer.cpp
//
// Copyright (c) 2026, Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
// SPDX-License-Identifier: MPL-2.0
//
/**
 * \file
 * \brief Write an example OSI MCAP trace file with metadata and channels.
 */

#include <osi-utilities/tracefile/TraceFileConfig.h>
#include <osi-utilities/tracefile/writer/MCAPTraceFileWriter.h>

#include <filesystem>
#ifdef _WIN32
#include <process.h>
#else
#include <unistd.h>
#endif

#include "osi_sensorview.pb.h"
#include "osi_version.pb.h"

/**
 * \brief Generate a temporary output path for the example MCAP file.
 * \return Temporary file path.
 */
auto GenerateTempFilePath() -> std::filesystem::path {
#ifdef _WIN32
    const auto pid = std::to_string(_getpid());
#else
    const auto pid = std::to_string(getpid());
#endif
    return std::filesystem::temp_directory_path() / ("sv_example_" + pid + ".mcap");  // add sv to indicate sensor view as recommended by the OSI-specification
}

/**
 * \brief Entry point for the MCAP writer example.
 */
auto main(int /*argc*/, const char** /*argv*/) -> int {
    std::cout << "Starting MCAP Writer example:" << std::endl;

    // Create writer and open file
    auto trace_file_writer = osi3::MCAPTraceFileWriter();
    const auto trace_file_path = GenerateTempFilePath();
    std::cout << "Creating trace_file at " << trace_file_path.string() << std::endl;

    mcap::McapWriterOptions mcap_options("osi");
    // Use the library default chunk size (16 MiB). Lichtblick plays back well
    // with 4-32 MiB chunks; adjust via kMinChunkSize / kMaxChunkSize bounds.
    mcap_options.chunkSize = osi3::tracefile::config::kDefaultChunkSize;
    // Default: zstd
    mcap_options.compression = mcap::Compression::Lz4;

    trace_file_writer.Open(trace_file_path, mcap_options);

    // add required and optional metadata to the net.asam.osi.trace metadata record
    auto net_asam_osi_trace_metadata = osi3::MCAPTraceFileWriter::PrepareRequiredFileMetadata();
    // Add optional metadata to the net.asam.osi.trace metadata record, as recommended by the OSI specification.
    net_asam_osi_trace_metadata.metadata["description"] = "Example mcap trace file created with the ASAM OSI utilities library.";  // optional field
    net_asam_osi_trace_metadata.metadata["creation_time"] = osi3::MCAPTraceFileWriter::GetCurrentTimeAsString();                   // optional field
    net_asam_osi_trace_metadata.metadata["authors"] = "Jane Doe, John Doe";                                                        // optional field
    if (!trace_file_writer.AddFileMetadata(net_asam_osi_trace_metadata)) {
        std::cerr << "Failed to add required metadata to trace_file." << std::endl;
        exit(1);
    }

    // add a channel to store some data
    const std::string topic = "Sensor_1_Input";
    const std::unordered_map<std::string, std::string> channel_metadata = {
        {"net.asam.osi.trace.channel.description", "This channel contains the input data (SensorView) for sensor 1"}};
    trace_file_writer.AddChannel(topic, osi3::SensorView::descriptor(), channel_metadata);

    // create OSI data to store
    const auto osi_version = osi3::InterfaceVersion::descriptor()->file()->options().GetExtension(osi3::current_interface_version);

    osi3::SensorView sensor_view_1;
    sensor_view_1.mutable_version()->CopyFrom(osi_version);
    sensor_view_1.mutable_sensor_id()->set_value(0);
    sensor_view_1.mutable_host_vehicle_id()->set_value(12);

    auto* const ground_truth_1 = sensor_view_1.mutable_global_ground_truth();
    ground_truth_1->mutable_version()->CopyFrom(osi_version);

    auto* const host_vehicle = ground_truth_1->mutable_moving_object()->Add();
    host_vehicle->mutable_id()->set_value(12);
    host_vehicle->mutable_vehicle_classification()->set_type(osi3::MovingObject_VehicleClassification_Type_TYPE_SMALL_CAR);
    host_vehicle->mutable_base()->mutable_dimension()->set_length(5);
    host_vehicle->mutable_base()->mutable_dimension()->set_width(2);
    host_vehicle->mutable_base()->mutable_dimension()->set_height(1.5);
    host_vehicle->mutable_base()->mutable_velocity()->set_x(10.0);

    // write the data continuously in a loop
    constexpr double kTimeStepSizeS = 0.1;  // NOLINT
    for (int i = 0; i < 10; ++i) {
        // manipulate the data so not every message is the same
        constexpr auto kNsPerSec = osi3::tracefile::config::kNanosecondsPerSecond;
        auto timestamp = sensor_view_1.timestamp().seconds() * kNsPerSec + sensor_view_1.timestamp().nanos();
        timestamp += kTimeStepSizeS * kNsPerSec;
        sensor_view_1.mutable_timestamp()->set_nanos(timestamp % kNsPerSec);
        sensor_view_1.mutable_timestamp()->set_seconds(static_cast<int64_t>(timestamp / kNsPerSec));
        ground_truth_1->mutable_timestamp()->set_nanos(timestamp % kNsPerSec);
        ground_truth_1->mutable_timestamp()->set_seconds(static_cast<int64_t>(timestamp / kNsPerSec));
        const auto old_position = host_vehicle->base().position().x();
        const auto new_position = old_position + host_vehicle->base().velocity().x() + kTimeStepSizeS;
        host_vehicle->mutable_base()->mutable_position()->set_x(new_position);

        // finally write the data using topic
        trace_file_writer.WriteMessage(sensor_view_1, topic);
    }

    // Close the file to flush the last piece of data from memory and ensure a clean exit state.
    trace_file_writer.Close();

    std::cout << "Finished MCAP Writer example" << std::endl;
}

File: python/examples/example_mcap_writer.py
# SPDX-License-Identifier: MPL-2.0
# SPDX-FileCopyrightText: Copyright (c) 2026, Bayerische Motoren Werke Aktiengesellschaft (BMW AG)

"""Write an example OSI MCAP trace file with metadata and channels."""

from __future__ import annotations

import os
import sys
from datetime import datetime, timezone
from pathlib import Path

import google.protobuf
from osi3.osi_sensorview_pb2 import SensorView

from osi_utilities import MultiTraceWriter
from osi_utilities.timestamp import timestamp_to_nanoseconds
from osi_utilities.tracefile import DEFAULT_CHUNK_SIZE, MAX_CHUNK_SIZE, MIN_CHUNK_SIZE


def _generate_osi_filename(description: str, extension: str, frame_count: int = 10) -> str:
    """Generate filename following OSI naming convention."""
    now = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    try:
        from osi3.osi_version_pb2 import DESCRIPTOR as VERSION_DESCRIPTOR
        from osi3.osi_version_pb2 import InterfaceVersion

        osi_ver = VERSION_DESCRIPTOR.file_options.Extensions[InterfaceVersion.current_interface_version]
        osi_version = f"{osi_ver.version_major}.{osi_ver.version_minor}.{osi_ver.version_patch}"
    except Exception:
        osi_version = "0.0.0"
    proto_version = google.protobuf.__version__
    pid = os.getpid()
    return f"{now}_sv_{osi_version}_{proto_version}_{frame_count}_{pid}_{description}.{extension}"


def main() -> int:
    """Create a SensorView MCAP file with 10 frames of a moving vehicle."""
    print("Starting MCAP Writer example:")

    output_dir = Path(__file__).resolve().parent.parent.parent / ".playground"
    output_dir.mkdir(exist_ok=True)
    trace_file_path = output_dir / _generate_osi_filename("example_mcap_writer", "mcap")
    print(f"Creating trace file at {trace_file_path}")
    print(
        f"  Config: chunk_size={DEFAULT_CHUNK_SIZE // (1024 * 1024)} MiB"
        f" (range: {MIN_CHUNK_SIZE // (1024 * 1024)}-{MAX_CHUNK_SIZE // (1024 * 1024)} MiB)"
    )

    metadata = {
        "description": "Example MCAP trace file created with the ASAM OSI utilities Python SDK.",
    }

    writer = MultiTraceWriter()
    if not writer.open(trace_file_path, metadata=metadata):
        print(f"Error: Could not open '{trace_file_path}'", file=sys.stderr)
        return 1

    topic = "Sensor_1_Input"
    writer.add_channel(
        topic,
        SensorView,
        metadata={"net.asam.osi.trace.channel.description": "Input data (SensorView) for sensor 1"},
    )

    # Build a SensorView with one moving object
    sensor_view = SensorView()
    sensor_view.sensor_id.value = 0
    sensor_view.host_vehicle_id.value = 12
    gt = sensor_view.global_ground_truth
    host = gt.moving_object.add()
    host.id.value = 12
    host.vehicle_classification.type = 2  # TYPE_SMALL_CAR
    host.base.dimension.length = 5
    host.base.dimension.width = 2
    host.base.dimension.height = 1.5
    host.base.velocity.x = 10.0

    time_step_ns = int(0.1 * 1e9)
    with writer:
        for i in range(10):
            ts_ns = timestamp_to_nanoseconds(sensor_view) + time_step_ns
            sensor_view.timestamp.seconds = int(ts_ns // 1e9)
            sensor_view.timestamp.nanos = int(ts_ns % 1e9)
            gt.timestamp.CopyFrom(sensor_view.timestamp)

            host.base.position.x += host.base.velocity.x * 0.1

            if not writer.write_message(sensor_view, topic):
                print(f"Error: Could not write message {i}", file=sys.stderr)
                return 1

        # Add custom metadata using the name + dict overload (alternative to constructor metadata)
        writer.add_file_metadata(
            "custom.example.metadata",
            {"source_tool": "example_mcap_writer", "scenario": "straight_road_acceleration"},
        )

    print("Finished MCAP Writer example")
    return 0


if __name__ == "__main__":
    sys.exit(main())
